src/main/java/picard/analysis/CollectQualityYieldMetrics.java

```diff
@@ -87,64 +89,98 @@
+        // of bases if there are secondary alignments in the input file.
+        private final boolean includeSecondaryAlignments;
 
-        final byte[] quals;
-        if (USE_ORIGINAL_QUALITIES) {
-            byte[] tmp = rec.getOriginalBaseQualities();
-            if (tmp == null) tmp = rec.getBaseQualities();
-            quals = tmp;
-        } else {
-            quals = rec.getBaseQualities();
+        // If true, include bases from supplemental alignments in metrics. Setting to true may cause double-counting
+        // of bases if there are supplemental alignments in the input file.
+        public final boolean includeSupplementalAlignments;
+
+        // The metrics to be accumulated
+        private final QualityYieldMetrics metrics = new QualityYieldMetrics();
+
+        public QualityYieldMetricsCollector(final boolean useOriginalQualities,
+                                            final boolean includeSecondaryAlignments,
+                                            final boolean includeSupplementalAlignments) {
+            this.useOriginalQualities          = useOriginalQualities;
+            this.includeSecondaryAlignments    = includeSecondaryAlignments;
+            this.includeSupplementalAlignments = includeSupplementalAlignments;
         }
 
-        // add up quals, and quals >= 20
-        for (final int qual : quals) {
-            metrics.Q20_EQUIVALENT_YIELD += qual;
-            if (qual >= 20) metrics.Q20_BASES++;
-            if (qual >= 30) metrics.Q30_BASES++;
+        public void acceptRecord(final SAMRecord rec, final ReferenceSequence ref) {
+            if (!this.includeSecondaryAlignments    && rec.getNotPrimaryAlignmentFlag()) return;
+            if (!this.includeSupplementalAlignments && rec.getSupplementaryAlignmentFlag()) return;
+
+            final int length = rec.getReadLength();
+            metrics.TOTAL_READS++;
+            metrics.TOTAL_BASES += length;
 
+            final boolean isPfRead = !rec.getReadFailsVendorQualityCheckFlag();
             if (isPfRead) {
-                metrics.PF_Q20_EQUIVALENT_YIELD += qual;
-                if (qual >= 20) metrics.PF_Q20_BASES++;
-                if (qual >= 30) metrics.PF_Q30_BASES++;
+                metrics.PF_READS++;
+                metrics.PF_BASES += length;
+            }
+
+            final byte[] quals;
+            if (this.useOriginalQualities) {
+                byte[] tmp = rec.getOriginalBaseQualities();
+                if (tmp == null) tmp = rec.getBaseQualities();
+                quals = tmp;
+            } else {
+                quals = rec.getBaseQualities();
+            }
+
+            // add up quals, and quals >= 20
+            for (final int qual : quals) {
+                metrics.Q20_EQUIVALENT_YIELD += qual;
+                if (qual >= 20) metrics.Q20_BASES++;
+                if (qual >= 30) metrics.Q30_BASES++;
```

Comment:
```
if(qual >= 30) {
    metrics.Q20_BASES++;
    metrics.Q30_BASES++;
} else if (qual >=20) {
    metrics.Q20_BASES++;
}
```

slightly faster for qualities above 30 not sure if it makes it less readable though